test/utils/Fuzzers.sol

```diff
@@ -158,4 +162,152 @@ abstract contract Fuzzers is Constants, Utils {
+            tranches[i].amount = tranchesWithTimestamps[i].amount;
+        }
+    }
+
+    /// @dev Fuzzes the tranche amounts and calculates the create amounts (total, deposit, protocol fee and broker
+    /// fee).
+    function fuzzTranchedStreamAmounts(
+        uint128 upperBound,
+        LockupTranched.Tranche[] memory tranches,
+        UD60x18 protocolFee,
+        UD60x18 brokerFee
+    )
+        internal
+        pure
+        returns (uint128 totalAmount, Lockup.CreateAmounts memory createAmounts)
+    {
+        uint256 trancheCount = tranches.length;
+        uint128 maxTrancheAmount = upperBound / uint128(trancheCount * 2);
+
+        // Precompute the first tranche amount to prevent zero deposit amounts.
+        tranches[0].amount = boundUint128(tranches[0].amount, 100, maxTrancheAmount);
+        uint128 estimatedDepositAmount = tranches[0].amount;
+
+        // Fuzz the other tranche amounts by bounding from 0.
+        unchecked {
+            for (uint256 i = 1; i < trancheCount; ++i) {
+                uint128 trancheAmount = boundUint128(tranches[i].amount, 0, maxTrancheAmount);
+                tranches[i].amount = trancheAmount;
+                estimatedDepositAmount += trancheAmount;
+            }
+        }
+
+        // Calculate the total amount from the approximated deposit amount (recall that the sum of all tranche amounts
+        // must equal the deposit amount) using this formula:
+        //
+        // $$
+        // total = \frac{deposit}{1e18 - protocolFee - brokerFee}
+        // $$
+        totalAmount = ud(estimatedDepositAmount).div(ud(uUNIT - protocolFee.intoUint256() - brokerFee.intoUint256()))
+            .intoUint128();
+
+        // Calculate the fee amounts.
+        createAmounts.protocolFee = ud(totalAmount).mul(protocolFee).intoUint128();
+        createAmounts.brokerFee = ud(totalAmount).mul(brokerFee).intoUint128();
+
+        // Here, we account for rounding errors and adjust the estimated deposit amount and the tranches. We know
+        // that the estimated deposit amount is not greater than the adjusted deposit amount below, because the inverse
+        // of {Helpers.checkAndCalculateFees} over-expresses the weight of the fees.
+        createAmounts.deposit = totalAmount - createAmounts.protocolFee - createAmounts.brokerFee;
+        tranches[tranches.length - 1].amount += (createAmounts.deposit - estimatedDepositAmount);
+    }
+
+    /// @dev Fuzzes the durations.
+    function fuzzTrancheDurations(LockupTranched.TrancheWithDuration[] memory tranches) internal pure {
+        unchecked {
+            // Precompute the first tranche duration.
+            tranches[0].duration = uint40(_bound(tranches[0].duration, 1, 100));
+
+            // Bound the durations so that none is zero and the calculations don't overflow.
+            uint256 durationCount = tranches.length;
+            uint40 maxDuration = (MAX_UNIX_TIMESTAMP - tranches[0].duration) / uint40(durationCount);
```

Comment:
because in the `fuzzTrancheTimestamps` function, we calculate the fuzzed `timestamp` by summing the `startTime` and `i * step` - we need to have one less length due to adding the `startTime`; otherwise, it would be greater than `MAX_UNIX_TIMESTAMP`

https://github.com/sablier-labs/v2-core/blob/b94cd4da0150fd78f44b74e621a2c598241fba5b/test/utils/Fuzzers.sol#L306

Reply:
1. wrt `fuzzTrancheTimestamps()`, the timestamp of the last tranche may still overflow.

```solidity
R1: start = tranches[0].timestamp;
R2: step = (MAX_UNIX_TIMESTAMP - tranches[0].timestamp) / (trancheCount - 1);
R3: halfStep = step / 2;
for (uint256 i = 1; i < trancheCount; ++i) {
    R4: timestamp = start + i * step;
    R5: _bound(timestamp, timestamp - halfStep, timestamp + halfStep);
}
```

  Based on the above relations, the max possible value for timestamp of the last tranche ($index:= n-1$) could be

```solidity
start + (trancheCount - 1) * step + halfStep
```

$\implies {ts}_{n-1} = ts_0 + (n - 1) \cdot \frac{{\text{{max}} - ts_0}}{{n-1}} + \frac{{\text{{max}} - ts_0}}{{n-1}} \cdot \frac{1}{2}$

$\implies {ts}_{n-1} = \text{{max}} + \frac{{\text{{max}} - ts_0}}{{n-1}} \cdot \frac{1}{2}$

$\implies {ts}_{n-1} > max$

That's why my point that If the goal is to keep well below `MAX_UNIX_TIMESTAMP` then we should divide with `trancheCount` in `fuzzTrancheTimestamps()` for calculating step. In that case:

$\implies {ts}_{n-1} = \text{{max}} - \frac{{\text{{max}} - t_0}}{{2n}} < max$

-------------
2. I agree that in `fuzzTrancheDurations()` calculations, we should divide by `durationCount` to keep the timestamp of last tranche below `MAX_UNIX_TIMESTAMP`. 

    But currently, `maxDuration` is calculated as a difference of `MAX_UNIX_TIMESTAMP` and `tranches[0].duration` which can still compute the timestamp of last tranche above `MAX_UNIX_TIMESTAMP`. Therefore, `maxDuration` should be calculated as the difference of `MAX_UNIX_TIMESTAMP` and `block.timestamp`.

```solidity
uint40 maxDuration = (MAX_UNIX_TIMESTAMP - getBlockTimestamp()) / uint40(durationCount);
```

In other, the intent is to divide the total period between `MAX_UNIX_TIMESTAMP` and the startTime which is `getBlockTimestamp()`.

Reply:
i think you are missing the fact `MAX_UNIX_TIMESTAMP` is $2^{32}$ and not $2^{40}$, so it won't overflow in Solidity:

https://github.com/sablier-labs/v2-core/blob/b94cd4da0150fd78f44b74e621a2c598241fba5b/test/utils/Constants.sol#L9

See [here](https://en.wikipedia.org/wiki/Year_2038_problem#:~:text=Consequently%2C%20if%20a%20signed%2032,on%20Tuesday%2C%2019%20January%202038.) why we've set it to this value.

Reply:
> But currently, maxDuration is calculated as a difference of MAX_UNIX_TIMESTAMP and tranches[0].duration which can still compute the timestamp of last tranche above MAX_UNIX_TIMESTAMP. Therefore, maxDuration should be calculated as the difference of MAX_UNIX_TIMESTAMP and block.timestamp.

I finally understood 😆

Reply:
> i think you are missing the fact `MAX_UNIX_TIMESTAMP` is $2^{32}$ and not $2^{40}$, so it won't overflow in Solidity.

My bad. I didn't mean overflow. I wanted to say that the timestamp of the last tranche may still be above `MAX_UNIX_TIMESTAMP`. Is that acceptable to us?